### ISSUE TITLE
Add more user-friendly way to add user accounts

### ISSUE THREAD

**Issue by czlee** [TabbycatDebate/tabbycat#502]
Now that, after five years, we have the luxury to be able to think about such niceties…

What I'm guessing we _probably_ want is a single form that bulk-invites many users, using (ideally) just:
- e-mail address
- whether they should have a tab director account or an assistant account

The user wouldn't be created upon invitation. Instead, it would e-mail each of them a link to a sign-up form, which creates an account upon submission. But in our case, they shouldn't be allowed to sign up _unless_ they've received an invitation. This means the signup form needs to be controlled with access tokens, and you shouldn't be able to access the signup form without it.

The other use-case is the one we assumed would be best back in 2014, before we started setting up sites a few days before the tournament: where you're physically next to the tab assistants and just have them create accounts while logged into your account, one by one. This workflow would basically be the same as how you create accounts in Django admin, with a less intimidating UI.

Django's built-in authentication system doesn't support this. There are probably third-party apps that do, but I think our use-case is a little niche, so the apps don't seem to cover exactly what we want. For example:
- https://github.com/pennersr/django-allauth/
- https://github.com/bee-keeper/django-invitations/

**Comment by philipbelesky** [comment]
The inability of tabbykitten copies to access email complicates this. I wonder also if the two use cases could be combined, in that:

- There's a front end form to create a new user account (I don't think a bulk-add form is necessary; people can just open with tabs if need be)
- That form prompts people to set a password
- Submitting a form creates the user as normal
- If the ability to send email is present it sends an email to the user noting their account creation, along with a link to the change password form.

This seems like it would cover the two general use cases:

- Adding assistants 'in the room' in which case they can enter their own passwords and have no need to confirm the account
- Adding accounts for tab directors or other parties prior to the tournament, whereby their password could by initially set, but then easily changed by the invited user.

What was the reasoning for not creating a user upon 'invitation'? It seems like it creates a lot of hassle around auth tokens for not much upside? In the worst case having an extraneous user account hanging around doesn't seem like a downside.

**Comment by czlee** [comment]
I can't say I particularly care if this feature isn't available in Tabbykitten. It's a naturally limited version of the software, we discourage people from using it wherever possible and I don't think we should be restricting what we implement just because it complicates Tabbykitten.

The bulk-add is just so that you can add (say) five to ten users by typing in their e-mail addresses and then hitting "submit", without having to click through forms five to ten times. It only makes sense in a context where you want to create accounts prior to the tournament.

The reason you don't create a user upon invitation is because you typically want them to be able to set their username. You need auth tokens either way—best practice is to force users to set their password at first login. In the status quo you couldn't do this by forcing a "reset password" e-mail to be sent to them, which will just feel a bit odd because they didn't forget their password, but will basically be the same thing.

Having thought about this more, I'm no longer convinced that any of this (in any form) should be a priority at all, unless there's evidence that making user accounts is unduly holding up users.

**Comment by philipbelesky** [comment]
We should probably get some stats on how popular the kitten version is.
While it’s limited it’s not limited in a particularly important manner: the
main distinguishing feature is just being able to email people.

I don’t think ts particularly necessary to have users set their own
username, I’ve only ever just used first names and never had to change
them.

Agree it’s not a priority, although the simple version of this I outlined
is probably easily achieved.

**Comment by philipbelesky** [comment]
A lot of this work (I think) has been done in that 'blank site start' flow, where users are prompted to create a new account if they don't have one already. So any new page for adding additional users could draw from that.

**Comment by tfpk** [comment]
I think the simplest way to implement this is as follows:
- on account creation, set two "passwords" for a tournament, one for assistants and one for admins.
- allow anyone to access an "account creation" page (or maybe just let the page exist, with no link on the site)
 - require anyone who wants to use that account creation page to also use one of those passwords.
 - the password they use determines the type of account they have.

The tournament convenor/tab director could then email or fb message anyone who needed an account the relevant password. They'd be able to create an account as normal.

Pros:
 - Easy to implement
 - No need for an inbuilt email (so works with tabby-cat)
 - Works over fb/in person/myspace/carrier pigeon

Cons:
 - Not as secure as an individual link (though anyone who has an account on tabbycat could just share their account details, and we still get names/emails, so I don't think this is a big concern).

**Comment by czlee** [comment]
I thought about this a bit more, and realized that the email part I suggested in my original comment isn't really essential, the key thing is that users have a unique link with access token (typically embedded in the url). I think this is mostly analogous to password resets, actually: notice that you can _almost_ do all this now with password resets, it's just that you can't do it in bulk, and users can't choose their username. So this simpler system would be similar, except that instead of users having a single-use access token each, we'd have (a) a single access token that everyone uses (per account level), and (b) it'd be taken in a password field rather than the URL.

I'd prefer not (b), but the main argument there is just user convenience (distributing a link is easier than distributing a link and a password), not security. I think (a) is a good idea, since you can then distribute this in a group chat or something (with people you presumably trust, and you can review what accounts exist later). I think I'd just be a bit more comfortable about security if either this link expired after some fixed time (say, 24 hours), or if while the link was enabled, some persistent reminder message showed up somewhere reminding the tab director that they probably want to disable it at some point?

**Comment by tfpk** [comment]
In regards to (b), I think it's just as simple to have the key in the link as a GET argument, and then the page can fill that in - it's simpler than doing the magic all in the URL, I think? And in regards to (a), I was definitely thinking that there'd be a boolean "allow account signups" and a warning banner or something letting people know they should turn that off before the start of the tournament.

**Comment by czlee** [comment]
Django's framework actually makes it very marginally easier to do it in the URL—URL arguments are matched in `self.kwargs`, which you can just look up; GET arguments are in `self.request.GET`. But they're both pretty easy. For comparison, Django's authentication app does password reset links in the URL. 

Re (a), sounds good!

**Comment by tfpk** [comment]
I've hit a bit of a decision point I want to run past you both. Basically, Django's `User` is unique to an instance, but as far as I can see we can have multiple tournaments going, to which a single `User` object has access. Because that's the case, I can't see a good place to put the key for user registration, or where to put preferences about it. There are basically four options I see:

a) Use existing tournament objects, and have it so each tournament has a token for assistants and directors, with the toggle for whether the token works that we discussed above. This is the simplest, but it also means that you can sign up with one tournament's keys and have it work for another one. You might also end up with multiple sets of keys that give the new user the same properties.
b) Do the above, but force users to be unique to a tournament. This fixes the issue, but  is a lot of work, especially when we have to account for some users who are able to edit multiple/all tournaments. It's also functionality I definitely don't think is called for - I don't see a good use case for it.
c) Create new infrastructure for "tabby cat instance settings" (which I don't think exists at the moment) where you can set the keys for every tournament in the instance. Would work, but again it's a lot of infrastructure and I'm not sure exactly what it would be used for apart from this.
d) Do (a), but enforce uniformity between the keys (so any instance can change the settings for every other instance).
I think I prefer d, but it's going to be kind of messy. What are your opinions? It's quite possible I'm missing something obvious.

**Comment by czlee** [comment]
Ah, this is an excellent question. Regarding (c), we don't currently have infrastructure for site-wide preferences, but it's not because we don't need one. It's just that there are only about two site-wide preferences in existence, and they're both defunct (they're in "league options", which are not documented or officially supported).

It's not unthinkable that we'll need more, so personally I wouldn't be opposed to creating a new infrastructure for it. In fact, the preferences library we use ([django-dynamic-preferences](https://django-dynamic-preferences.readthedocs.io/en/latest/)) supports global preferences, and they're a bit easier to manage than per-instance preferences. So it's in principle possible, and might be able to leverage existing infrastructure for per-tournament preferences, if appropriately abstracted/decomposed.

I agree with your assessment of (b). I'd rather not tie these to tournaments as in (a), for the reason you describe.

As for (d), I think there's a way to do this more akin to (c) than to (a). You set up a view using `TournamentMixin`, but the only thing "tournament" about it is the nav bar. That is, rather than trying to enforce uniformity between what is essentially duplicate, you just have a bunch of views that are essentially the same except for the URL and nav bar, and that all edit the same data.

Thoughts?
